### ISSUE TITLE
Bug in integer promotion/conversion/multiplication

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/GAP.jl#332]
Problem is caused by `Int64` values `>= 2^60`, which cannot be stored in a small int.

```
julia>  2^60 * GAP.Globals.Factorial(50)
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `*' on 2 arguments
ERROR: Error thrown by GAP
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] error_handler() at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/GAP.jl:50
 [3] #call_gap_func#5(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(GAP.call_gap_func), ::Main.ForeignGAP.MPtr, ::Int64, ::Vararg{Any,N} where N) at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/ccalls.jl:149
 [4] call_gap_func(::Main.ForeignGAP.MPtr, ::Int64, ::Vararg{Any,N} where N) at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/ccalls.jl:139
 [5] #call#7(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Main.ForeignGAP.MPtr, ::Int64, ::Vararg{Any,N} where N) at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/GAP.jl:136
 [6] (::Main.ForeignGAP.MPtr)(::Int64, ::Vararg{Any,N} where N) at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/GAP.jl:136
 [7] *(::Int64, ::Main.ForeignGAP.MPtr) at /Users/mhorn/Projekte/OSCAR/GAP.jl/src/adapter.jl:69
 [8] top-level
```

**Comment by fingolfin** [comment]
Really annoying is that right now we can't even avoid this by using `julia_to_gap`:
```
julia> GAP.julia_to_gap(2^60) * GAP.Globals.Factorial(50)
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `*' on 2 arguments
ERROR: Error thrown by GAP
```

Only thing that works is to first convert to a BigInteger or Int128:
```
julia> GAP.julia_to_gap(Int128(2^60)) * GAP.Globals.Factorial(50)
GAP: 35065062095372263678204347284693182438521523255008459599785630170611712000000000000

julia> GAP.julia_to_gap(big(2^60)) * GAP.Globals.Factorial(50)
GAP: 35065062095372263678204347284693182438521523255008459599785630170611712000000000000
```

The problem is caused by `Int64` values `>= 2^60`, which cannot be stored in a small int. What happens is that this value then is wrapped in a `T_JULIA` GAP object, which then is passed to GAP's multiplication operation. Which runs into an error, because it doesn't know how to multiple a `T_JULIA` with a GAP object.

One way to resolve this is to convert *all* `Int64` to native GAP integers (so those not fitting into a GAP small integer are turned into a GAP larger integer); and conversely, GAP larger integers that fit into an Int64 are converted to `Int64`, period.


Note that there are still issues; it's confusing that the same function sometimes returns an `Int64` and sometimes an `MPtr` for which `isinteger` does not return `true`, but which still can used like an integer in many situations, but not always... And this also means that seemingly similar operations can have very different behavior. E.g. of the following two similar code snippets, one produces garbage
```
julia> (@gap Factorial(19))^2
-8682740722073337856

julia> (@gap Factorial(20))^2
GAP: 5919012181389927685417441689600000000

```

This highlights that this package is not really meant to be used by "regular users"; instead it's a sharp tool that experts can use to build other packages which then can be exposed to users.

Of course we could *also* provide an end-user ready API in this package, but for now, this is not a priority, I'd say.

**Comment by fingolfin** [comment]
This was resolved by #357
